src/solver.jl

```diff
@@ -9,79 +10,106 @@
 
 
 # Policy struct created according to one used in DiscreteValueIteration
+
+#TODO: Dosctring
+struct FiniteHorizonSolver <: Solver
+    verbose::Bool
+end
+
+#TODO: Dosctring
 mutable struct FiniteHorizonValuePolicy{Q<:AbstractArray, U<:AbstractMatrix, P<:AbstractMatrix, A, M<:MDP} <: Policy
     qmat::Q
     util::U 
     policy::P 
     action_map::Vector{A}
     include_Q::Bool 
-    mdp::M
+    m::M
 end
 
 # Policy constructor
-function FiniteHorizonValuePolicy(mdp::MDP)
-    return FiniteHorizonValuePolicy(zeros(mdp.horizon, mdp.no_states, length(mdp.actions)), zeros(mdp.horizon, mdp.no_states), ones(Int64, mdp.horizon, mdp.no_states), ordered_actions(mdp), true, mdp)
+function FiniteHorizonValuePolicy(m::MDP)
+    return FiniteHorizonValuePolicy(zeros(m.horizon + 1, length(stage_states(m, 1)), length(actions(m))), zeros(m.horizon + 1, length(stage_states(m, 1))), ones(Int64, m.horizon + 1, length(stage_states(m, 1))), ordered_actions(m), true, m)
 end
 
 function action(policy::FiniteHorizonValuePolicy, s::S) where S
-    sidx = stage_stateindex(policy.mdp, s, s.epoch)
+    sidx = stateindex(policy.m, s)
```

Comment:
With my MDP, this is causing issues. `stateindex` is using the "global" mdp indexing while `stage_stateindex` is using a "local" stage indexing. Because the solution algorithm works on the stage indexes, fetching action based on `stateindex` from the "global" indexing will get a wrong action. 

So this line should imo be:
`sidx = stage_stateindex(policy.mdp, s, stage(s))`

Maybe a 2D testacase with custom MDPs would help with this?

Reply:
Just a note, changing to `sidx = stage_stateindex(policy.mdp, s, stage(s))` fixes my problem in https://github.com/JuliaPOMDP/FiniteHorizonPOMDPs.jl/pull/8#issuecomment-775271381, but I am not sure if it's not causing problems elsewhere.

Reply:
> Just a note, changing to `sidx = stage_stateindex(policy.mdp, s, stage(s))` fixes my problem in [#8 (comment)](https://github.com/JuliaPOMDP/FiniteHorizonPOMDPs.jl/pull/8#issuecomment-775271381), but I am not sure if it's not causing problems elsewhere.

I strongly think we should NOT have the three argument `stage_stateindex(mdp, s, stage)`. The stage information should be included in the state.

Reply:
I think the current `stateindex` is the correct solution. The `policy.policy` stores policy for each state, so you also need a global index of state.
`policy.policy` is `horizon x (no_states + 1)` big. Getting local index would only obtain actions from first row/column.
Please provide example.

Reply:
> > Just a note, changing to `sidx = stage_stateindex(policy.mdp, s, stage(s))` fixes my problem in [#8 (comment)](https://github.com/JuliaPOMDP/FiniteHorizonPOMDPs.jl/pull/8#issuecomment-775271381), but I am not sure if it's not causing problems elsewhere.
> 
> I strongly think we should NOT have the three argument `stage_stateindex(mdp, s, stage)`. The stage information should be included in the state.

@zsunberg I believe this corresponds to the structure suggested in former Readme.md file and the method itself was implemented by you. While I agree this is somewhat redundant information, I think this supports the idea that the method returns the local index for the given stage, thus supports readability.

Reply:
> I believe this corresponds to the structure suggested in former Readme.md file

It could be :smile: If that was there, I have changed my mind.

> While I agree this is somewhat redundant information, I think this supports the idea that the method returns the local index for the given stage, thus supports readability.

In my opinion he difference in name between `stateindex` and `stage_stateindex` makes this clear.

I think having the redundant information could lead to confusion. For example, what should happen if
```julia
stage_stateindex(m, s, 2)
```
if `stage(s)` is `3`. I.e. what should happen if the third argument disagrees with the second?

Reply:
(I am open to arguments for doing it the other way though)

Reply:
I think that if the user is able to do `stage_stateindex(m, s, 2)` then he could be also able to define all states to a specific stage, say 2. So it depends on the point of view.
That said I understand your point and will change this.

Reply:
I am in favor of 2-argument `stage_stateindex(m, s)` calling `stage(s)` internally as well.

I was trying to make a different point, I will try to better explain what I think might be an issue here.
My point here is that `policy` is defined as a matrix:
```
mutable struct FiniteHorizonValuePolicy{Q<:AbstractArray, U<:AbstractMatrix, P<:AbstractMatrix, A, M<:MDP} <: Policy
    qmat::Q
    util::U 
    policy::P 
    action_map::Vector{A}
    include_Q::Bool 
    m::M
end
```

with rows holding the policy for each stage:
```
function addstagerecord(fhpolicy::FiniteHorizonValuePolicy, qmat, util, policy, stage)    
    fhpolicy.qmat[stage, :, :] = qmat
    fhpolicy.util[stage, :] = util
    fhpolicy.policy[stage, :] = policy
    return fhpolicy
end
```

However, the `action` method accessed policy with scalar index (stateindex returns Integer, not a CartesianIndex):
```
function action(policy::FiniteHorizonValuePolicy, s::S) where S
    sidx = stateindex(policy.m, s)
    aidx = policy.policy'[sidx]
    return policy.action_map[aidx]
end
```

Indexing with scalalar into matrices works and I think it would work fine in MDPs where `stateindex` transforms to `stage_stateindex` easily with mod n_stages, but it will fail if the user-defined `stateindex` orders states differently. 


I see now that my recommendation was wrong, but I still don't think that the current code works correctly. I think that the new code for `action` should be
```
function action(policy::FiniteHorizonValuePolicy, s::S) where S
    sidx = stage_stateindex(policy.m, s)
    aidx = policy.policy'[stage(s), sidx]
    return policy.action_map[aidx]
end
```

I think the reason the current tests don't catch this is either because the stateindex->stage_stateindex wraps nicely with mod or because in 1-d gridworld, the optimal policy doesn't contain a whole lot of different actions so the mismatched indexes still return the correct action.

PS: wouldn't it be better to store the policy in columns?

Reply:
Also, I think it would be nice to have a test MDP instance with different number of states in each stage (and different transition probs as well) to catch weird edge cases.

Reply:
In my opinion, the `[stage(), stage_stateindex]` should work the same as `[stateindex]`.

Column-based store: I have to yet study this feature of Julia, but this is to be done.
A different number of states in each stage: This feature is desirable. However, it is not possible right now. I am creating an issue for that.

Given that each stage would have a different number of states, we would have to also update `FHPolicy`, because it initializes variables as matrices. With the different number of states, we would have to initialize them as `(horizon(m) + 1) x max_states_in_stage` or as another structure, presumably as some sort of dictionary.

Reply:
> In my opinion, the [stage(), stage_stateindex] should work the same as [stateindex].

I think this works only when `stateindex(s) = (stage(s)-1)*length(stage_states) + stage_stateindex`, i.e. when the stateindex wraps "nicely" at the end of the rows. I assume that is why transpose of the matrix before indexing is needed, the scalar indexing works by columns, and the code wouldn't work with `'` removed.

`stateindex` is up to user to define. So for example, consider MDP with 3 states `s, s', s''`  and in stage 1 they had `state_stageindex` `1,2,3,` their `stateindex` could be completeley independent of the `state_stageindex`, for example defined as `4, 1, 13`. Indexing with `stateindex` would then fetch actions for states in different stages. 

Let me know if I misunderstood how the indexing works here, it really seems to me that while the current code works with the existing test cases, it should be modified to accommodate less specialized user definitions of `stateindex`. I am not sure wheter `stateindex()` is used elsewhere, but it should be used carefully as user implementations may order states differently than `state_stageindex()`.

Otherwise I think the code looks great!

Reply:
You are right.
I already changed the code, as this is more general.